core/src/main/java/com/scalar/db/storage/dynamo/DynamoTableMetadataManager.java

```diff
@@ -127,19 +156,173 @@ private DataType convertDataType(String columnType) {
+                    .map(pKey -> AttributeValue.builder().s(pKey).build())
+                    .collect(Collectors.toList()))
+            .build());
+    itemValues.put(
+        CLUSTERING_KEY,
+        AttributeValue.builder()
+            .l(
+                metadata.getClusteringKeyNames().stream()
+                    .map(pKey -> AttributeValue.builder().s(pKey).build())
+                    .collect(Collectors.toList()))
+            .build());
+    itemValues.put(
+        SECONDARY_INDEX, AttributeValue.builder().ss(metadata.getSecondaryIndexNames()).build());
+
+    PutItemRequest request =
+        PutItemRequest.builder()
+            .tableName(getFullTableName(namespacePrefix, METADATA_NAMESPACE, METADATA_TABLE))
+            .item(itemValues)
+            .build();
+
+    try {
+      client.putItem(request);
+    } catch (DynamoDbException e) {
+      throw new StorageRuntimeException(
+          "adding meta data for table "
+              + getFullTableName(namespacePrefix, namespace, table)
+              + " failed",
+          e);
+    }
   }
 
   @Override
   public Set<String> getTableNames(String namespace) {
-    // TODO To implement
-    throw new UnsupportedOperationException();
+    Set<String> tableSet = new HashSet<>();
+    try {
+      ListTablesResponse listTablesResponse = client.listTables();
+      List<String> tableNames = listTablesResponse.tableNames();
+      for (String tableName : tableNames) {
+        if (tableName.startsWith(getFullNamespaceName(namespacePrefix, namespace))) {
+          tableSet.add(tableName);
+        }
+      }
+    } catch (DynamoDbException e) {
+      throw new StorageRuntimeException("retrieving the table names failed", e);
+    }
+    return tableSet;
+  }
+
+  private void createMetadataTableIfNotExist() throws StorageRuntimeException {
+    if (metadataTableExists()) {
+      return;
+    }
+
+    CreateTableRequest.Builder requestBuilder = CreateTableRequest.builder();
+    List<AttributeDefinition> columnsToAttributeDefinitions = new ArrayList<>();
+    columnsToAttributeDefinitions.add(
+        AttributeDefinition.builder()
+            .attributeName(TABLE)
+            .attributeType(ScalarAttributeType.S)
+            .build());
+    requestBuilder.attributeDefinitions(columnsToAttributeDefinitions);
+    requestBuilder.keySchema(
+        KeySchemaElement.builder().attributeName(TABLE).keyType(KeyType.HASH).build());
+    requestBuilder.provisionedThroughput(
+        ProvisionedThroughput.builder()
+            .readCapacityUnits(METADATA_TABLE_RU)
+            .writeCapacityUnits(METADATA_TABLE_RU)
+            .build());
+    requestBuilder.tableName(getFullTableName(namespacePrefix, METADATA_NAMESPACE, METADATA_TABLE));
+
+    try {
+      client.createTable(requestBuilder.build());
+    } catch (DynamoDbException e) {
+      throw new StorageRuntimeException("creating meta data table failed");
+    }
+
+    while (true) {
```

Comment:
This block is redundant with the block in Admin.
(CREATING_WAITING_TIME is also redundantly defined)
It would be great if the redundancy is removed.

Reply:
Yes, indeed. I created a static function for waiting table creation. Applied in [391f30b](https://github.com/scalar-labs/scalardb/commit/391f30b54026c24a765bd56b672570feca195ed3)